### DISCUSSION TITLE
What are the resource requirements to deploy the Kindling?

### DISCUSSION THREAD

**Discussion by HeyJocky** [harmonycloud/kindling#115] [General]
If I want to monitor 100 nodes, how many resources do I need to prepare to deploy the Kindling？ And 1000 nodes?

**Comment by dxsup** [comment]
We can't just tell how many resources you need to prepare based on the little information you provided. It depends on many factors. What we can provide is which part of Kindling consumes how many resources. Then you could calculate the total resources by yourself based on this information. 

There are three parts that consume resources for using Kindling: Kindling-agent, Prometheus, and Grafana.

## Kindling-agent
The agent should be deployed on each node, so the resources consumption I talk about later is based on per node. 

### Memory
You can simply use the deployment file we provide at `deploy/kindling-deploy.yml` to deploy the agent in your Kubernetes. We **limit the total memory usage to 1100mi** (300mi for kindling-probe and 800mi for kindling-collector, in case you are curious). It is typically enough to run the agent on a node **under a 10k QPS load**. 

But note that the memory usage is relevant with many factors, such as QPS, the complexity of your HTTP endpoints, the number of Pods, etc. The real usage could vary greatly on different nodes. I suggest users concentrate on the memory after deploying the agent and give it more if your load is much higher than 10k QPS. However, there is still a normal range for memory usage, which is 100mi~2000mi. If you find the usage of your agent exceeds this range or keeps increasing, then it is unreasonable and very likely a bug. Please don't hesitate to open an issue, and we will fix that as soon as possible. We are always on our way to profile the agent.

### CPU
We don't limit the CPU usage at the deployment file provided. But as both the Kindling-probe and the Kindling-collector could only use one thread for processing the main part of data for now, **the maximum usage is about 2500m**. If your load on one node is much higher than 15k, the agent will drop events and the final metrics may be inaccurate. We have had a plan on implementing multi-thread processing to increase the amount of data the agent can handle.

### Storage
Although Kindling-agent won't use storage to store data, we have to remind users that it could consume **at most 2.5GB to store its running logs by default**. You could change the maximum size of log files in the section `logger` of the configuration. Remember to check the storage is available before using Kindling.

## Prometheus
We use Prometheus to scrap and store our metrics. While Prometheus is a monitoring system, in both performance and operational terms it is a database. So it requests a lot of memory and storage if you want to gain good performance.

### CPU and Memory
It depends on how many timeseries there are to be written and to be queried. More timeseries there are, more CPU and memory is needed. There are not too many documents saying how to calculate the CPU or memory consumption, but you could find some at [How much RAM does Prometheus 2.x need for cardinality and ingestion?](https://www.robustperception.io/how-much-ram-does-prometheus-2-x-need-for-cardinality-and-ingestion)  and [Why does Prometheus use so much RAM?](https://www.robustperception.io/why-does-prometheus-use-so-much-ram). As you can see, it is complicated to calculate precisely the memory needed.

The number of timeseries Kindling-agent emits depends on many factors, such as the complexity of your HTTP endpoints, the number of Pods, the complexity of your network topology, and so on. But to be honest, we are unable to give the formula to calculate how many timeseries your agents will send for now. 

So if you want to know how many CPU and memory resources you should prepare, we can just give a combination value for reference only, and you have to adjust it based on your real environment. If you just want to experience the basic features Kindling provides in your development environment, **1 CPU and 1GB** is enough to start Prometheus. And from practice, it is highly recommended to prepare **at least 4 CPU and 16GB in your production environment** to optimize its performance. 

### Storage
Calculating the storage usage of Prometheus is tough work. It also depends on how many timeseries the agent emits. As what we have talked about in the last section, we are unable to tell how to calculate the number of timeseries. And we can't be irresponsible to give an arbitrary value of how much storage it will consume per day. When starting to use Kindling, we recommend users concentrate on the size of storage used after a week and record the size used per day, based on which a plan of capacity can be made as the later usages will be similar.

Now Kindling is developing increasingly in different ways. We are working hard on helping users use Kindling more easily and will provide a more detailed document on how to calculate the storage usage later. And also we would appreciate users contributing their plans of resources, which would help Kindling a lot.

## Grafana
Grafana is used to query and display the topology and the dashboards we provide. It won't increase too many resources if you have had one yourself, so no resources you need to adjust for Grafana. If you don't have one and you want to know how many resources are requested by Grafana, we recommend users prepare **at least 300MB memory and 1 CPU**, and you can also find the official recommendations from [here](https://grafana.com/docs/grafana/latest/installation/requirements/#hardware-recommendations).
